pkg/server/api.go

```diff
@@ -186,31 +189,123 @@ func (sh *APIHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	// for now, serve v2 if we receive a request without a useragent header with Ignition
-	// this happens if the user pings the endpoint directly (e.g. with curl)
-	// and we don't want to break existing behaviour
-	if !strings.HasPrefix(useragent, "Ignition/") {
-		return semver.New("2.2.0"), nil
-	}
-	ignVersionString := strings.SplitAfter(useragent, "/")[1]
-	ignSemver, err := semver.NewVersion(ignVersionString)
-	if err != nil {
-		return ignSemver, errors.Errorf("invalid version in useragent: %s", useragent)
+type acceptHeaderValue struct {
+	MIMEType    string
+	MIMESubtype string
+	Version     *semver.Version
+	QValue      float32
+}
+
+// Parse an accept header, ignoring any extensions that aren't
+// either version or relative quality factor q.
+func parseAcceptHeader(input string) ([]acceptHeaderValue, error) {
+	var header []acceptHeaderValue
+
+	values := strings.Split(input, ",")
+	for i, value := range values {
+		// remove spaces
+		values[i] = strings.TrimSpace(values[i])
+		parts := strings.Split(value, ";")
+		for j := range parts {
+			parts[j] = strings.TrimSpace(parts[j])
+		}
+
+		if !strings.Contains(parts[0], "/") {
+			// This is not a MIME type, ignore bad data
+			continue
+		}
+		// mtype[0] is the main MIME type, mtype[1] is the sub MIME type
+		mtype := strings.SplitN(parts[0], "/", 2)
+
+		// check value extensions for version and q parameters, ignore other extensions
+		var v *semver.Version
+		var q float32
+		versionIsSet := false
+		qIsSet := false
+		for _, ext := range parts[1:] {
+			if strings.Contains(ext, "=") {
+				keyval := strings.SplitN(ext, "=", 2)
+				if keyval[0] == "version" && !versionIsSet {
+					var err error
+					v, err = semver.NewVersion(keyval[1])
+					if err != nil {
+						// This is not a valid version
+						continue
+					}
+					versionIsSet = true
+				} else if keyval[0] == "q" && !qIsSet {
+					q64, err := strconv.ParseFloat(keyval[1], 32)
+					if err != nil {
+						// This is not a valid relative quality factor
+						continue
+					}
+					q = float32(q64)
+					qIsSet = true
+				}
+			}
+		}
+
+		// Default version to nil pointer
+		if !versionIsSet {
+			v = nil
+		}
+		// Default q to 1
+		if !qIsSet {
+			q = 1.0
+		}
+
+		header = append(header, acceptHeaderValue{
+			mtype[0],
+			mtype[1],
+			v,
+			q,
+		})
 	}
 
-	if !ignSemver.LessThan(*semver.New("0.22.0")) && ignSemver.LessThan(*semver.New("1.0.0")) {
-		// versions [0.22.0;1) support config spec v2.2, and
-		return semver.New("2.2.0"), nil
-	} else if !ignSemver.LessThan(*semver.New("2.3.0")) && ignSemver.LessThan(*semver.New("3.0.0")) {
-		// versions [2.3.0;3) support config spec v3.1
-		return semver.New("3.1.0"), nil
+	if len(header) == 0 {
+		return nil, errors.New("no valid accept header detected")
+	}
+
+	// sort headers by descending q factor value
+	sort.SliceStable(header, func(i, j int) bool { return header[i].QValue > header[j].QValue })
```

Comment:
question for my understanding : why sort?

Reply:
We want to select the acceptable Ignition version with the highest client `q` value, and the client isn't required to present them in order.

Reply:
thanks @bgilbert ! :smile_cat: